agentops_dcp.go

```diff
@@ -50,6 +66,117 @@ type VbSeqNoEntry struct {
 // GetVBucketSeqnosCallback is invoked with the results of `GetVBucketSeqnos` operations.
 type GetVBucketSeqnosCallback func([]VbSeqNoEntry, error)
 
+// OpenStream opens a DCP stream for a particular VBucket.
+func (agent *Agent) OpenCollectionStream(vbId uint16, flags DcpStreamAddFlag, vbUuid VbUuid, startSeqNo,
+	endSeqNo, snapStartSeqNo, snapEndSeqNo SeqNo, evtHandler CollectionStreamObserver, filter *CollectionStreamFilter, cb OpenStreamCallback) (PendingOp, error) {
+	var req *memdQRequest
+	handler := func(resp *memdQResponse, _ *memdQRequest, err error) {
+		if resp != nil && resp.Magic == resMagic {
+			// This is the response to the open stream request.
+			if err != nil {
+				req.Cancel()
+
+				// All client errors are handled by the StreamObserver
+				cb(nil, err)
+				return
+			}
+
+			numEntries := len(resp.Value) / 16
+			entries := make([]FailoverEntry, numEntries)
+			for i := 0; i < numEntries; i++ {
+				entries[i] = FailoverEntry{
+					VbUuid: VbUuid(binary.BigEndian.Uint64(resp.Value[i*16+0:])),
+					SeqNo:  SeqNo(binary.BigEndian.Uint64(resp.Value[i*16+8:])),
+				}
+			}
+
+			cb(entries, nil)
+			return
+		}
+
+		if err != nil {
+			req.Cancel()
+			evtHandler.End(vbId, err)
+			return
+		}
+
+		// This is one of the stream events
+		switch resp.Opcode {
+		case cmdDcpSnapshotMarker:
+			vbId := uint16(resp.Vbucket)
+			newStartSeqNo := binary.BigEndian.Uint64(resp.Extras[0:])
+			newEndSeqNo := binary.BigEndian.Uint64(resp.Extras[8:])
+			snapshotType := binary.BigEndian.Uint32(resp.Extras[16:])
+			evtHandler.SnapshotMarker(newStartSeqNo, newEndSeqNo, vbId, SnapshotState(snapshotType))
+		case cmdDcpMutation:
+			vbId := uint16(resp.Vbucket)
+			seqNo := binary.BigEndian.Uint64(resp.Extras[0:])
+			revNo := binary.BigEndian.Uint64(resp.Extras[8:])
+			flags := binary.BigEndian.Uint32(resp.Extras[16:])
+			expiry := binary.BigEndian.Uint32(resp.Extras[20:])
+			lockTime := binary.BigEndian.Uint32(resp.Extras[24:])
+			evtHandler.Mutation(seqNo, revNo, flags, expiry, lockTime, resp.Cas, resp.Datatype, vbId, resp.Key, resp.Value)
```

Comment:
resp.Key contains both the collection id and the document key, I'm unsure if we should split these or leave together (as has been here). Interested in your opinion?